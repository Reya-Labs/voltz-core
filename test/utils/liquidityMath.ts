import { expect } from "../shared/expect";
import { LiquidityMathTest } from "../../typechain/LiquidityMathTest";
import { ethers, waffle } from "hardhat";
import snapshotGasCost from "../shared/snapshotGasCost";

const { BigNumber } = ethers;

describe("LiquidityMath", () => {
  let liquidityMath: LiquidityMathTest;
  const fixture = async () => {
    const factory = await ethers.getContractFactory("LiquidityMathTest");
    return (await factory.deploy()) as LiquidityMathTest;
  };
  beforeEach("deploy LiquidityMathTest", async () => {
    liquidityMath = await waffle.loadFixture(fixture);
  });

  describe("#addDelta", () => {
    it("1 + 0", async () => {
      expect(await liquidityMath.addDelta(1, 0)).to.eq(1);
    });
    it("1 + -1", async () => {
      expect(await liquidityMath.addDelta(1, -1)).to.eq(0);
    });
    it("1 + 1", async () => {
      expect(await liquidityMath.addDelta(1, 1)).to.eq(2);
    });
    it("2**128-15 + 15 overflows", async () => {
      await expect(
        liquidityMath.addDelta(BigNumber.from(2).pow(128).sub(15), 15)
      ).to.be.reverted;
    });
    it("0 + -1 underflows", async () => {
      await expect(liquidityMath.addDelta(0, -1)).to.be.reverted;
    });
    it("3 + -4 underflows", async () => {
      await expect(liquidityMath.addDelta(3, -4)).to.be.reverted;
    });
    it("gas add", async () => {
      await snapshotGasCost(liquidityMath.getGasCostOfAddDelta(15, 4));
    });
    it("gas sub", async () => {
      await snapshotGasCost(liquidityMath.getGasCostOfAddDelta(15, -4));
    });
  });
});
