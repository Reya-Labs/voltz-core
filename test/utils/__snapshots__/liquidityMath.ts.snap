// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiquidityMath #addDelta gas add 1`] = `268`;

exports[`LiquidityMath #addDelta gas sub 1`] = `367`;
